src/connector/src/sink/dynamodb.rs

```diff
@@ -184,76 +214,95 @@ impl DynamoDbRequest {
+        &mut self,
+        key: HashMap<String, AttributeValue>,
+        request_items: &mut Vec<DynamoDbRequest>,
+    ) {
         let key = key
             .into_iter()
             .filter(|(k, _)| self.dynamodb_keys.contains(k))
             .collect();
         let del_req = DeleteRequest::builder().set_key(Some(key)).build().unwrap();
         let req = WriteRequest::builder().delete_request(del_req).build();
-        self.write_one_req(req);
+        self.write_one_req(req, request_items);
     }
 
-    fn write_one_req(&mut self, req: WriteRequest) {
+    fn write_one_req(&mut self, req: WriteRequest, request_items: &mut Vec<DynamoDbRequest>) {
         let r_req = DynamoDbRequest {
             inner: req,
             key_items: self.dynamodb_keys.clone(),
         };
         if let Some(v) = r_req.extract_pk_values() {
-            self.request_items.retain(|item| {
+            request_items.retain(|item| {
                 !item
                     .extract_pk_values()
                     .unwrap_or_default()
                     .iter()
                     .all(|x| v.contains(x))
             });
         }
-        self.request_items.push(r_req);
+        request_items.push(r_req);
     }
 
-    async fn write_chunk(&mut self) -> Result<()> {
-        if !self.request_items.is_empty() {
-            let table = self.table.clone();
-            let req_items = std::mem::take(&mut self.request_items)
-                .into_iter()
-                .map(|r| r.inner)
-                .collect();
-            let reqs = hashmap! {
-                table => req_items,
-            };
-            self.client
-                .batch_write_item()
-                .set_request_items(Some(reqs))
-                .return_consumed_capacity(ReturnConsumedCapacity::None)
-                .return_item_collection_metrics(ReturnItemCollectionMetrics::None)
-                .send()
-                .await
-                .map_err(|e| {
-                    SinkError::DynamoDb(
-                        anyhow!(e).context("failed to delete item from DynamoDB sink"),
-                    )
-                })?;
-        }
+    fn write_chunk(
+        &mut self,
+        request_items: Vec<DynamoDbRequest>,
+    ) -> Result<
+        Vec<
+            impl Future<
+                Output = result::Result<
+                    BatchWriteItemOutput,
+                    SdkError<BatchWriteItemError, HttpResponse>,
+                >,
+            >,
+        >,
+    > {
+        let table = self.table.clone();
+        let req_items: Vec<Vec<_>> = request_items
+            .into_iter()
+            .map(|r| r.inner)
+            .chunks(self.max_batch_item_nums)
+            .into_iter()
+            .map(|chunk| chunk.collect())
+            .collect();
+        let futures: Vec<_> = req_items
+            .into_iter()
+            .map(|req_items| {
+                let reqs = hashmap! {
+                    table.clone() => req_items,
```

Comment:
We can return a `try_join_all` future directly in this method, and we can create the `try_join_all` future from an iterator directly to avoid two unnecessary `collect` and subsequent `into_iter` in this method.